python/ml4ir/applications/ranking/features/feature_fns/categorical.py

```diff
@@ -47,12 +49,22 @@ def __init__(self, feature_info: dict, file_io: FileIO, **kwargs):
                 currently supports either embedding or one_hot
             embedding_size : int
                 dimension size of categorical embedding
+            embedding_initializer: string
+                the tensorflow initializer to use for the embedding matrix
         """
         super().__init__(feature_info=feature_info, file_io=file_io, **kwargs)
 
-        self.vocabulary_keys, _ = get_vocabulary_info(self.feature_layer_args,
-                                                      self.file_io,
-                                                      self.default_value)
+        self.vocabulary = self.feature_layer_args.get(self.VOCABULARY)
+        if isinstance(self.vocabulary, str):
+            self.feature_layer_args["vocabulary_file"] = self.vocabulary
+            self.vocabulary_keys, _ = get_vocabulary_info({VocabularyInfoArgsKey.VOCABULARY_FILE: self.vocabulary},
+                                                          self.file_io,
+                                                          self.default_value)
+        elif isinstance(self.vocabulary, list):
+            self.vocabulary_keys = self.vocabulary
+        else:
+            raise NotImplementedError("Unsupported value for argument vocabulary")
+
```

Comment:
Adding support for passing either a vocabulary file (as earlier) or a list of strings (will be used to generate query type embeddings)

Reply:
This layer is just a categorical vectorizer - either onehot or embedding. Check the unchanged code below for context.